sys/net/gnrc/link_layer/lwmac/rx_state_machine.c

```diff
@@ -0,0 +1,349 @@
+            lwmac_packet_info_t info;
+            int ret = _parse_packet(pkt, &info);
+
+            if(ret != 0) {
+                LOG_DEBUG("Packet could not be parsed: %i\n", ret);
+                gnrc_pktbuf_release(pkt);
+                continue;
+            }
+
+            if(info.header->type == FRAMETYPE_BROADCAST) {
+                _dispatch_defer(lwmac->rx.dispatch_buffer, pkt);
+                continue;
+            }
+
+            /* TODO:
+             * If we see a WA here we have a rough clue about the wakeup phase
+             * of this node. But there is no timestamping of incoming frames yet
+             * so maybe add a timestamp to every frame in event callback. */
+
+            if(info.header->type != FRAMETYPE_WR) {
+                LOG_DEBUG("Packet is not WR: 0x%02x\n", info.header->type);
+                gnrc_pktbuf_release(pkt);
+                continue;
+            }
+
+            /* No need to keep pkt anymore */
+            gnrc_pktbuf_release(pkt);
+
+            if(!_addr_match(&lwmac->l2_addr, &info.dst_addr)) {
+                LOG_DEBUG("Packet is WR but not for us\n");
+                continue;
+            }
+
+            /* Save source address for later addressing */
+            lwmac->rx.l2_addr = info.src_addr;
+
+            found_wr = true;
+            break;
+        }
+
+        if(!found_wr) {
+            LOG_DEBUG("No WR found, stop RX\n");
+            GOTO_RX_STATE(RX_STATE_FAILED, true);
+        }
+
+        // TODO: don't flush queue
+        packet_queue_flush(&lwmac->rx.queue);
+
+        GOTO_RX_STATE(RX_STATE_SEND_WA, true);
+    }
+    case RX_STATE_SEND_WA:
+    {
+        LOG_DEBUG("RX_STATE_SEND_WA\n");
+
+        gnrc_pktsnip_t* pkt;
+        gnrc_netif_hdr_t* nethdr_wa;
+
+        assert(lwmac->rx.l2_addr.len != 0);
+
+        /* Assemble WA packet */
+        lwmac_frame_wa_t lwmac_hdr;
+        lwmac_hdr.header.type = FRAMETYPE_WA;
+        lwmac_hdr.dst_addr = lwmac->rx.l2_addr;
+
+        pkt = gnrc_pktbuf_add(NULL, &lwmac_hdr, sizeof(lwmac_hdr), GNRC_NETTYPE_LWMAC);
+        if(pkt == NULL) {
+            LOG_ERROR("Cannot allocate pktbuf of type GNRC_NETTYPE_LWMAC\n");
+            GOTO_RX_STATE(RX_STATE_FAILED, true);
+        }
+
+        pkt = gnrc_pktbuf_add(pkt, NULL, sizeof(gnrc_netif_hdr_t) + lwmac->rx.l2_addr.len, GNRC_NETTYPE_NETIF);
+        if(pkt == NULL) {
+            LOG_ERROR("Cannot allocate pktbuf of type GNRC_NETTYPE_NETIF\n");
+            GOTO_RX_STATE(RX_STATE_FAILED, true);
+        }
+
+        /* We wouldn't get here if add the NETIF header had failed, so no
+           sanity checks needed */
+        nethdr_wa = (gnrc_netif_hdr_t*) _gnrc_pktbuf_find(pkt, GNRC_NETTYPE_NETIF);
+
+        /* Construct NETIF header and insert address for WA packet */
+        gnrc_netif_hdr_init(nethdr_wa, 0, lwmac->rx.l2_addr.len);
+        gnrc_netif_hdr_set_dst_addr(nethdr_wa, lwmac->rx.l2_addr.addr, lwmac->rx.l2_addr.len);
```

Comment:
Do we still need to set the dst-address of the netif header for WA packet here? Since WA will be broadcasted, its dst-address in the MAC header (or, netif header) will be ```0xFFFF```. LW-MAC put the dst-address in the ```lwmac_frame_wa_t``` for WA packet, right?  :-)